Optimizer: Improve performance of large InlineArrays at Onone

eeckstein opened the pull request:
We need to avoid copying large InlineArrays at Onone, because even for Onone the performance suffers too much if such copies are not eliminated.

The solution to this to run the TempRValueElimination at Onone.
In this PR I added a new MandatoryTempRValueElimination pass, which works as the original TempRValueElimination, except that it does not remove any alloc_stack instruction which are associated with source variables.

In addition it's required to simplify `mark_dependence` instructions at Onone. I re-implement the SILCombine peephole as a Swift instruction simplification and let run this simplification also in the OnoneSimplification pass.

rdar://151629149

eeckstein committed:
SIL: improve and fix mark-dependence instruction APIs

* Move the mutating APIs into Context.swift, because SIL can only be mutated through a MutatingContext
* move the `baseOperand` and `base` properties from the instruction classes to the `MarkDependenceInstruction` protocol
* add `valueOrAddressOperand` and `valueOrAddress` in the `MarkDependenceInstruction` protocol

eeckstein committed:
Optimizer: do `mark_dependence` simplification at Onone

* re-implement the SILCombine peephole as a Swift instruction simplification
* run this simplification also in the OnoneSimplification pass

eeckstein committed:
Optimizer: run TempRValueElimination also at Onone

Introduce a new pass MandatoryTempRValueElimination, which works as the original TempRValueElimination, except that it does not remove any alloc_stack instruction which are associated with source variables.

Running this pass at Onone helps to reduce copies of large structs, e.g. InlineArrays or structs containing InlineArrays.
Copying large structs can be a performance problem, even at Onone.

rdar://151629149

nate-chandler reviewed (approved):
LGTM

atrick reviewed (commented):
LGTM but there should be a comment at the top of SimplifyMarkDependence that it cannot run before dependency diagnostics. Dependencies on trivial values are meaningful for diagnostics.

eeckstein committed:
AliasAnalysis: workaround a bug in the move-only checker

This is a workaround for a bug in the move-only checker: rdar://151841926.
The move-only checker sometimes inserts destroy_addr within read-only static access scopes.
Therefore don't consider static access scopes as immutable scopes.

eeckstein commented:
@swift-ci clean apple silicon benchmark

eeckstein commented:
@swift-ci benchmark

eeckstein commented:
@swift-ci apple silicon benchmark

eeckstein commented:
@swift-ci test macos
